inferno-core/src/Inferno/Infer.hs

```diff
@@ -988,13 +990,14 @@ infer expr =
               ImplType isMerged $ head ts_res,
               (Set.fromList ics)
                 `Set.union` cs_e
+                `Set.union` (Set.unions patConstraints)
                 `Set.union` patTysEqConstraints
                 `Set.union` patTysMustEqCaseExprTy t_e
                 `Set.union` patExpTysEqConstraints (zip (map (\(_, ty, _) -> ty) res) (map (\(_, _p, _, e'') -> e'') patExprs))
                 `Set.union` (Set.unions cs_res)
             )
           where
-            mkPatConstraint :: Pat (Pinned VCObjectHash) SourcePos -> Infer (InfernoType, [(Ident, TypeMetadata TCScheme)])
+            mkPatConstraint :: Pat (Pinned VCObjectHash) SourcePos -> Infer (InfernoType, [(Ident, TypeMetadata TCScheme)], Set.Set Constraint)
             mkPatConstraint pattern =
```

Comment:
I guess this PR doesn't introduce the `pattern` thing, but I'd still recommend changing it, especially as some modules define pattern synonyms